src/webpage/src/app.rs

```diff
@@ -0,0 +1,334 @@
+                if let ewebsock::WsEvent::Message(message) = message {
+                    if let ewebsock::WsMessage::Text(message) = message {
+                        let Ok(message_json) =
+                            serde_json::from_str::<serde_json::Value>(&message)
+                        else {
+                            continue;
+                        };
+                        let system_id: u8 =
+                            message_json["header"]["system_id"].as_u64().unwrap() as u8;
+                        let component_id: u8 =
+                            message_json["header"]["component_id"].as_u64().unwrap() as u8;
+                        let message_name: String =
+                            message_json["message"]["type"].as_str().unwrap().to_string();
+
+                        self.vehicles
+                            .entry(system_id)
+                            .or_insert_with(Default::default)
+                            .entry(component_id)
+                            .or_insert_with(Default::default);
+
+                        let messages = self
+                            .vehicles
+                            .get_mut(&system_id)
+                            .unwrap()
+                            .get_mut(&component_id)
+                            .unwrap();
+
+                        let message_info = messages
+                            .entry(message_name.clone())
+                            .or_insert_with(|| MessageInfo {
+                                last_sample_time: now,
+                                fields: Default::default(),
+                            });
+
+                        message_info.last_sample_time = now;
+
+                        if let Some(fields) = message_json["message"].as_object() {
+                            for (field_name, value) in fields {
+                                if field_name == "type" {
+                                    continue;
+                                }
+                                if let Some(num) = extract_number(value) {
+                                    let field_info = message_info
+                                        .fields
+                                        .entry(field_name.clone())
+                                        .or_insert_with(|| FieldInfo {
+                                            history: Vec::new(),
+                                            latest_value: num,
+                                        });
+                                    field_info.latest_value = num;
+                                    field_info.history.push((now, num));
+                                    if field_info.history.len() > 1000 {
+                                        field_info.history.remove(0);
+                                    }
+                                }
+                            }
+                        }
```

Comment:
```suggestion
                    let Ok(message_json) = serde_json::from_str::<serde_json::Value>(&message)
                    else {
                        continue;
                    };
                    let Some(system_id) = message_json["header"]["system_id"]
                        .as_u64()
                        .map(|n| n as u8)
                    else {
                        continue;
                    };
                    let Some(component_id) = message_json["header"]["component_id"]
                        .as_u64()
                        .map(|n| n as u8)
                    else {
                        continue;
                    };
                    let Some(message_name) = message_json["message"]["type"]
                        .as_str()
                        .map(|s| s.to_string())
                    else {
                        continue;
                    };

                    self.vehicles
                        .entry(system_id)
                        .or_default()
                        .entry(component_id)
                        .or_default();

                    let Some(vehicle) = self.vehicles.get_mut(&system_id) else {
                        continue;
                    };
                    let Some(messages) = vehicle.get_mut(&component_id) else {
                        continue;
                    };

                    let message_info =
                        messages.entry(message_name).or_insert_with(|| MessageInfo {
                            last_sample_time: now,
                            fields: Default::default(),
                        });

                    message_info.last_sample_time = now;

                    let Some(fields) = message_json["message"].as_object() else {
                        continue;
                    };

                    for (field_name, value) in fields {
                        if field_name == "type" {
                            continue;
                        }

                        let Some(num) = extract_number(value) else {
                            continue;
                        };

                        let field_info = message_info
                            .fields
                            .entry(field_name.clone())
                            .or_insert_with(|| FieldInfo {
                                history: Vec::new(),
                                latest_value: num,
                            });
                        field_info.latest_value = num;
                        field_info.history.push((now, num));
                        if field_info.history.len() > 1000 {
                            field_info.history.remove(0);
                        }
                    }
```

- combine if let matches
- remove the unwraps by using let else pattern
- fix formatting

Reply:
done.